clients/unshaded/src/main/java/tachyon/client/block/BufferedBlockInStream.java

```diff
@@ -0,0 +1,218 @@
+      return toRead;
+    }
+
+    if (toRead > mBuffer.limit() / 2) { // directly read if request is > one-half buffer size
+      mBufferPos = INVALID_BUFFER_POS; // Invalidate the buffer
+      int bytesRead = directRead(b, off, toRead);
+      mPos += bytesRead;
+      return bytesRead;
+    }
+
+    // For a read <= half the buffer size, fill the buffer first, then read from the buffer.
+    updateBuffer();
+    mBuffer.get(b, off, toRead);
+    mPos += toRead;
+    mBufferPos += toRead;
+    return toRead;
+  }
+
+  @Override
+  public long remaining() {
+    return mBlockSize - mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(pos >= 0, "Seek position is negative: " + pos);
+    Preconditions.checkArgument(pos <= mBlockSize, "Seek position is past end of block: "
+        + mBlockSize);
+    mBufferPos = INVALID_BUFFER_POS;
+    mPos = pos;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    checkIfClosed();
+    if (n <= 0) {
+      return 0;
+    }
+
+    long toSkip = Math.min(remaining(), n);
+    mBufferPos = INVALID_BUFFER_POS;
+    mPos += toSkip;
+    return toSkip;
+  }
+
+  protected long remainingInBuffer() {
+    if (mBufferPos == INVALID_BUFFER_POS) {
+      return 0;
+    }
+
+    return mBuffer.remaining();
+  }
+
+  /**
+   * Initializes the internal buffer based on the user's specified size. Any reads above half
+   * this size will not be buffered.
+   *
+   * @return a heap buffer of user configured size
+   */
+  private ByteBuffer allocateBuffer() {
+    TachyonConf conf = ClientContext.getConf();
+    return ByteBuffer.allocate((int) conf.getBytes(Constants.USER_REMOTE_READ_BUFFER_SIZE_BYTE));
+  }
+
+  /**
+   * Convenience method to ensure the stream is not closed.
+   */
+  protected void checkIfClosed() {
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+  }
+
+  /**
+   * Directly reads data to the given byte array. The data will not go through the internal
+   * buffer. This method should not modify mPos but should update the any metrics collection for
+   * bytes read.
+   *
+   * @param b the byte array to write the data to
+   * @param off the offset in the array to write to
+   * @param len the length of data to write into the array
+   * @return the number of bytes successfully read
+   * @throws IOException if an error occurs reading the data
+   */
+  protected abstract int directRead(byte[] b, int off, int len) throws IOException;
+
+  /**
+   * Updates the buffer so it is ready to be read from. After calling this method, the buffer should
+   * be positioned at 0 and mBufferPos should be equal to mPos. This method should not modify mPos
+   * but should update the any metrics collection for bytes read.
+   *
+   * @throws IOException if an error occurs reading the data
+   */
+  protected abstract void updateBuffer() throws IOException;
```

Comment:
try to make `updateBuffer` more concrete, while you may keep `bufferedRead` an abstract method which is required subclasses to implement.